README
Calc Mosaic README


INTRODUCTION

    Calc Mosaic is a program to create mosaic art from images with
OpenOffice Calc.  This can be used to create cool stop motion movies.

    See HISTORY for the story behind Calc Mosaic.


COMPILATION

    You need JDK, Apache ANT and OpenOffice SDK in order to compile
this program.

    You need juh.jar, jurt.jar, ridl.jar and unoil.jar in order to
compile the source.  They can be found in OpenOffice SDK.

${OpenOffice Installation}/ure/share/java/juh.jar
${OpenOffice Installation}/ure/share/java/jurt.jar
${OpenOffice Installation}/ure/share/java/ridl.jar
${OpenOffice Installation}/basis3.4/program/classes/unoil.jar

    Just copy them into the lib/ directory and they will work.  Also
You may want to review the property value of ooo.dir.sdk-classes in
build.xml.

    To compile, simply type "ant compile".  Type "ant jar" will
create the jar file.  Type "ant help" for more information.

    Alternatively, there is a pre-compiled jar file here within
the distribution.


USAGE

    Type "java -jar calcmosaic.jar -help" to display a simple help.

    The snapshots of a video can be grabbed using ffmpeg:

ffmpeg -i video.mp4 -r <framerate> -sameq -y -f image2 img%04d.jpg

    Note that a spreadsheet document can only contain at most 256
spreadsheets.

    Save file first for the macro to work.
    
    Hit the start button at A1 of the first and last sheet to play
the sheets.

    You can set the break time (in milliseconds) between sheets at
B1 of the first sheet (default 100 ms).

    You can set the background music at C1 of the first sheet.  This
only works on Linux and you must install mplayer.  Relative file path
is relative to the document file.  Play once before your live show to
avoid the I/O delay with cache.

    If you found problem running on Windows X64, see Windows-x64.txt.

    Hope you like it! ^_*'

imacat ^_*'
imacat@mail.imacat.idv.tw
2012-08-24


NEWS
NEWS in calcmosaic

2.1.0
	* Better user interface processing:  Added an icon, shortcut keys,
	  button focus, enter key, the number of chosen files, etc.
	* Added information and WinJava64.reg for the problem running on
	  MS Windows 64-bit with 64-bit Java JRE.  See:
	  https://issues.apache.org/ooo/show_bug.cgi?id=121258

2.0.0
	* You can run calcmosaic.jar from the file manager.  An
	  option dialog was added to obtain the options with GUI.
	* You can now run in the 0.x mode that paint the mosaic sheet cell
	  by cell, for presentation, by using the -slow switch.
    * Better localized in Traditional Chinese.
	* Errors are better handled.  Error messages are shown in either
	  STDERR or message box automatically.  Errors reading from
	  non-image files are handled.
	* The end-of-run message will not block for the user input when
	  running from console.
	* An API to return mosaic color data is added.
	* Many minor fixes.

1.0.3
	* Fixed build.xml includeantruntime warning.
	* Changed case of thisComponent to ThisComponent.

1.0.2
	* Fixed Javadoc problems.

1.0.1
	* Removed unused algorithms to calculate the mosaic cell color.

1.0.0
	* Rewritten to be faster (about 8 times faster).  Most of
	  program is now OpenOffice BASIC.  The Java program actually
	  injects OpenOffice BASIC macros into the document and run it.
	  Unlike Java UNO, OpenOffice BASIC does not refresh the display
	  for each UNO command.  The execution is now a LOT faster.
	* Added the play script and the start buttons, to help playing
	  the spreadsheet stop-motion.

0.2.1
	* Added a maximum check for the number of images used.  This
	  users from errors after hours of work of conversion.
	* Added HISTORY for the story behind Calc Mosaic.
	* Added explanation for the FFMPEG command to use for grabbing
	  snapshots.

0.2.0
	* Added the ability to convert multiple images.
	* Added the ability to calculate the mosaic cell size
	  automatically.
	* Added command line options and help.
	* Changed the option mosaic cell size to maximum number of
	  rows to use, to make it easier (and more meaningful) to
	  the users.
	* Changed the algorithm to calculate the color of the mosaic
	  cell, to have sharper result on the edges.

0.1.2
	* 0.1.1 mistakenly packaged 0.1.0 along with 0.1.1, and this is
	  fixed.

0.1.1
	* License changed from GPLv3 to Apache License v2.

0.1.0
	* Initial release



src/tw/idv/imacat/calcmosaic/Options.java
/* Copyright (c) 2012 imacat
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* 
 * Options
 * 
 * Created on 2012-10-29
 * 
 * Copyright (c) 2012 imacat
 */

package tw.idv.imacat.calcmosaic;

import java.io.File;

/**
 * The options of Calc Mosaic.
 * 
 * @author <a href="mailto:imacat&#64;mail.imacat.idv.tw">imacat</a>
 * @version 0.0.1
 */
public class Options {
    
    /** The image files. */
    public File files[] = new File[0];;
    
    /** The number of rows to use. */
    public int numRows = 0;
    
    /** Whether we should run in slow motion (for presentation). */
    public boolean isSlow = false;
    
    /**
     * Creates a new set of options.
     * 
     * @param files   the image files
     * @param numRows the number of rows to use
     * @param isSlow  whether we should run in slow motion (for
     *                presentation)
     */
    public Options(File files[], int numRows, boolean isSlow) {
        this.files = files;
        this.numRows = numRows;
        this.isSlow = isSlow;
    }
}


src/tw/idv/imacat/calcmosaic/OpenOffice.java
/* Copyright (c) 2012 imacat
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * OpenOffice
 * 
 * Created on 2012-08-16, rewritten from CalcMosaic
 * 
 * Copyright (c) 2012 imacat
 */

package tw.idv.imacat.calcmosaic;

import java.util.ResourceBundle;

import com.sun.star.beans.PropertyValue;
import com.sun.star.beans.XPropertySet;
import com.sun.star.comp.helper.Bootstrap;
import com.sun.star.bridge.XUnoUrlResolver;
import com.sun.star.frame.XComponentLoader;
import com.sun.star.frame.XDesktop;
import com.sun.star.lang.XComponent;
import com.sun.star.lang.XMultiComponentFactory;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XComponentContext;

/**
 * An OpenOffice instance.
 * 
 * @author <a href="mailto:imacat&#64;mail.imacat.idv.tw">imacat</a>
 * @version 2.0.0
 */
public class OpenOffice {
    
    /** The localization resources. */
    private ResourceBundle l10n = null;
    
    /** The desktop service. */
    private XDesktop desktop = null;
    
    /** The bootstrap context. */
    private XComponentContext bootstrapContext = null;
    
    /** The registry service manager. */
    private XMultiComponentFactory serviceManager = null;
    
    /**
     * Creates a new instance of Office.
     * 
     * @throws com.sun.star.comp.helper.BootstrapException if fails to
     *         create the initial component context
     */
    public OpenOffice()
            throws com.sun.star.comp.helper.BootstrapException {
        // Obtains the localization resources
        this.l10n = ResourceBundle.getBundle(
            this.getClass().getPackage().getName() + ".res.L10n");
        try {
            this.connect();
        } catch (com.sun.star.comp.helper.BootstrapException e) {
            throw e;
        }
    }
    
    /**
     * Connects to the OpenOffice program.
     * 
     * @throws com.sun.star.comp.helper.BootstrapException if fails to
     *         create the initial component context
     */
    private void connect()
            throws com.sun.star.comp.helper.BootstrapException {
        XMultiComponentFactory localServiceManager = null;
        Object unoUrlResolver = null;
        XUnoUrlResolver xUnoUrlResolver = null;
        Object service = null;
        
        if (this.isConnected()) {
            return;
        }
        
        // Obtains the local context
        try {
            this.bootstrapContext = Bootstrap.bootstrap();
        } catch (java.lang.Exception e) {
            throw new com.sun.star.comp.helper.BootstrapException(e);
        }
        if (this.bootstrapContext == null) {
            throw new com.sun.star.comp.helper.BootstrapException(
                this._("err_ooo_no_lcc"));
        }
        
        // Obtains the local service manager
        this.serviceManager = this.bootstrapContext.getServiceManager();
        
        // Obtain the desktop service
        try {
            service = this.serviceManager.createInstanceWithContext(
                "com.sun.star.frame.Desktop", this.bootstrapContext);
        } catch (com.sun.star.uno.Exception e) {
            throw new java.lang.UnsupportedOperationException(e);
        }
        this.desktop = (XDesktop) UnoRuntime.queryInterface(
            XDesktop.class, service);
        
        // Obtains the service manager
        this.serviceManager = this.bootstrapContext.getServiceManager();
        return;
    }
    
    /**
     * Returns whether the connection is on and alive.
     * 
     * @return true if the connection is alive, false otherwise
     */
    private boolean isConnected() {
        if (this.serviceManager == null) {
            return false;
        }
        try {
            UnoRuntime.queryInterface(
                XPropertySet.class, this.serviceManager);
        } catch (com.sun.star.lang.DisposedException e) {
            this.serviceManager = null;
            return false;
        }
        return true;
    }
    
    /**
     * Starts a new spreadsheet document.
     * 
     * @return the new spreadsheet document
     */
    public SpreadsheetDocument startNewSpreadsheetDocument() {
        XComponentLoader xComponentLoader = (XComponentLoader)
            UnoRuntime.queryInterface(
            XComponentLoader.class, this.desktop);
        PropertyValue props[] = new PropertyValue[0];
        final String url = "private:factory/scalc";
        XComponent xComponent = null;
        
        // Load a new document
        try {
            xComponent = xComponentLoader.loadComponentFromURL(url,
                "_default", 0, props);
        } catch (com.sun.star.io.IOException e) {
            throw new java.lang.RuntimeException(
                String.format(_("err_open"), url), e);
        } catch (com.sun.star.lang.IllegalArgumentException e) {
            throw new java.lang.RuntimeException(
                String.format(_("err_open"), url), e);
        }
        
        return new SpreadsheetDocument(xComponent);
    }
    
    /**
     * Gets a string for the given key from this resource bundle
     * or one of its parents.  If the key is missing, returns
     * the key itself.
     * 
     * @param key the key for the desired string 
     * @return the string for the given key 
     */
    private String _(String key) {
        try {
            return new String(
                this.l10n.getString(key).getBytes("ISO-8859-1"),
                "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            return this.l10n.getString(key);
        } catch (java.util.MissingResourceException e) {
            return key;
        }
    }
}


src/tw/idv/imacat/calcmosaic/NotAnImageException.java
/* Copyright (c) 2012 imacat
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * NotAnImageException
 * 
 * Created on 2012-10-27, rewritten from CalcMosaic
 * 
 * Copyright (c) 2012 imacat
 */

package tw.idv.imacat.calcmosaic;

/**
 * An exception thrown when the file read is not an image.
 * 
 * @author <a href="mailto:imacat&#64;mail.imacat.idv.tw">imacat</a>
 * @version 0.0.1
 */
public class NotAnImageException extends Exception {
    
    /**
     * Constructs a <code>NotAnImageException</code> exception with
     * <code>null</code> as its error detail message.
     * 
     */
    public NotAnImageException() {
        super();
    }
    
    /**
     * Constructs a <code>NotAnImageException</code> with the
     * specified detail message.
     * 
     * @param message The detail message (which is saved for later
     *                retrieval by the Throwable.getMessage() method)
     */
    public NotAnImageException(String message) {
        super(message);
    }
    
    /**
     * Constructs an <code>NotAnImageException</code> with the
     * specified detail message and cause. 
     * <p>
     * Note that the detail message associated with <code>cause</code>
     * is <em>not</em> automatically incorporated into this
     * exception's detail message.
     * 
     * @param message The detail message (which is saved for later
     *                retrieval by the Throwable.getMessage() method)
     * @param cause   The cause (which is saved for later retrieval
     *                by the Throwable.getCause() method). (A null
     *                value is permitted, and indicates that the
     *                cause is nonexistent or unknown.)
     */
    public NotAnImageException(String message, Throwable cause) {
        super(message, cause);
    }
    
    /**
     * Constructs an <code>NotAnImageException</code> with the
     * specified cause and a detail message of <code>(cause==null
     * &#63; null : cause.toString())</code> (which typically contains
     * the class and detail message of <code>cause</code>).  This
     * constructor is useful for IO exceptions that are little more
     * than wrappers for other throwables.
     * 
     * @param cause   The cause (which is saved for later retrieval
     *                by the Throwable.getCause() method). (A null
     *                value is permitted, and indicates that the
     *                cause is nonexistent or unknown.)
     */
    public NotAnImageException(Throwable cause) {
        super(cause);
    }
}


src/tw/idv/imacat/calcmosaic/OptionDialog.java
/* Copyright (c) 2012 imacat
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* 
 * OptionDialog
 * 
 * Created on 2012-10-29
 * 
 * Copyright (c) 2012 imacat
 */

package tw.idv.imacat.calcmosaic;

import java.io.File;
import java.util.ResourceBundle;

import java.awt.Dialog.ModalityType;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.GridLayout;
import java.awt.Toolkit;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JSpinner;
import javax.swing.SpinnerNumberModel;
import javax.swing.UIManager;

import javax.swing.JFileChooser;
import javax.swing.filechooser.FileNameExtensionFilter;

/**
 * An dialog to obtain the options.
 * 
 * @author <a href="mailto:imacat&#64;mail.imacat.idv.tw">imacat</a>
 * @version 2.1.0
 */
public class OptionDialog {
    
    /** The localization resources. */
    private ResourceBundle l10n = null;
    
    /** The option set. */
    private Options options = null;
    
    /** The dialog. */
    private JDialog dialog = new JDialog(
        null, ModalityType.TOOLKIT_MODAL);
    
    /** The image files. */
    public File files[] = new File[0];;
    
    /** The number of rows to use. */
    public int numRows = 0;
    
    /** Whether we should run in slow motion (for presentation). */
    public boolean isSlow = false;
    
    /** The current directory. */
    private File currentDirectory = null;
    
    /** The file chooser button. */
    private JButton filesButton = new JButton();
    
    /** The spinner of the number of rows. */
    private JSpinner numRowsSpinner = new JSpinner();
    
    /** The check box of whether we should run in slow motion. */
    private JCheckBox isSlowCheckBox = new JCheckBox();
    
    /** The OK button. */
    private JButton okButton = new JButton();
    
    /**
     * Creates a new option dialog.
     * 
     * @param files   the image files
     * @param numRows the number of rows to use
     * @param isSlow  whether we should run in slow motion (for
     *                presentation)
     */
    public OptionDialog(File files[], int numRows, boolean isSlow) {
        // Obtains the localization resources
        this.l10n = ResourceBundle.getBundle(
            this.getClass().getPackage().getName() + ".res.L10n");
        this.files = files;
        this.numRows = numRows;
        this.isSlow = isSlow;
    }
    
    /**
     * Asks the user and returns the options.
     * 
     * @return the options set by the user
     */
    public Options getOptions() {
        JPanel optionPanel = new JPanel();
        JLabel label = null;
        JButton cancelButton = new JButton(_("dialog_btn_cancel"));
        SpinnerNumberModel spinnerModel = new SpinnerNumberModel(
            this.numRows, 0, 32767, 1);
        Dimension screenSize
            = Toolkit.getDefaultToolkit().getScreenSize();
        Dimension frameSize = null;
        
        // Makes "enter" acts as "space" that activates the buttons
        UIManager.put("Button.defaultButtonFollowsFocus", true);
        
        this.dialog.setIconImage(CalcMosaic.icon);
        this.dialog.setTitle(String.format(_("dialog_opts_title"),
            CalcMosaic.APP_NAME));
        this.dialog.setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
        optionPanel.setLayout(new GridLayout(4, 2, 10, 5));
        
        // Reserves the space for a 3-digit number
        this.filesButton.setText(
            String.format(_("dialog_btn_files"), 256));
        this.filesButton.setMnemonic(KeyEvent.VK_F);
        this.filesButton.addActionListener(new ActionListener() {
                public void actionPerformed(ActionEvent e) {
                    chooseFiles();
                }
            });
        label = new JLabel(_("dialog_label_files"));
        label.setLabelFor(this.filesButton);
        optionPanel.add(label);
        optionPanel.add(this.filesButton);
        
        this.numRowsSpinner.setModel(spinnerModel);
        label = new JLabel(_("dialog_label_numrows"));
        label.setLabelFor(this.numRowsSpinner);
        optionPanel.add(label);
        optionPanel.add(this.numRowsSpinner);
        
        this.isSlowCheckBox.setText(_("dialog_checkbox_isslow"));
        this.isSlowCheckBox.setSelected(this.isSlow);
        label = new JLabel(_("dialog_label_isslow"));
        label.setLabelFor(this.isSlowCheckBox);
        optionPanel.add(label);
        optionPanel.add(this.isSlowCheckBox);
        
        this.okButton.setText(_("dialog_btn_ok"));
        this.okButton.setMnemonic(KeyEvent.VK_O);
        this.okButton.addActionListener(new ActionListener() {
                public void actionPerformed(ActionEvent e) {
                    options = new Options(files,
                        (Integer) numRowsSpinner.getValue(),
                        isSlowCheckBox.isSelected());
                    dialog.dispose();
                }
            });
        if (this.files.length == 0) {
            this.okButton.setEnabled(false);
        } else {
            this.okButton.setEnabled(true);
        }
        optionPanel.add(this.okButton);
        cancelButton = new JButton(_("dialog_btn_cancel"));
        cancelButton.setMnemonic(KeyEvent.VK_C);
        cancelButton.addActionListener(new ActionListener() {
                public void actionPerformed(ActionEvent e) {
                    options = null;
                    dialog.dispose();
                }
            });
        optionPanel.add(cancelButton);
        
        this.options = null;
        this.dialog.add(optionPanel);
        this.dialog.pack();
        frameSize = this.dialog.getSize();
        this.dialog.setLocation(
                (screenSize.width - frameSize.width) / 2,
                (screenSize.height - frameSize.height) / 2
            );
        this.filesButton.setText(
            String.format(_("dialog_btn_files"), this.files.length));
        this.dialog.setVisible(true);
        
        return this.options;
    }
    
    /**
     * Chooses the image files.
     * 
     */
    private void chooseFiles() {
        File chosen[] = new File[0];
        
        while (true) {
            JFileChooser chooser = new JFileChooser(
                this.currentDirectory);
            FileNameExtensionFilter filter = new FileNameExtensionFilter(
                _("dialog_file_chooser_image_type"), "jpg", "png", "gif");
            int result = -1;
            
            chooser.setDialogTitle(_("dialog_file_chooser_title"));
            chooser.setFileFilter(filter);
            chooser.setMultiSelectionEnabled(true);
            result = chooser.showOpenDialog(this.dialog);
            this.currentDirectory = chooser.getCurrentDirectory();
            
            if (result == JFileChooser.CANCEL_OPTION) {
                return;
            
            } else if (result == JFileChooser.ERROR_OPTION) {
                JOptionPane.showMessageDialog(this.dialog,
                    _("err_file_chooser_failed"));
                return;
            }
            
            chosen = chooser.getSelectedFiles();
            if (chosen.length > 256) {
                JOptionPane.showMessageDialog(this.dialog,
                    _("err_file_too_many"),
                    String.format(_("dialog_err_title"),
                        CalcMosaic.APP_NAME),
                    JOptionPane.ERROR_MESSAGE);
                continue;
            }
            break;
        }
        
        this.files = chosen;
        this.filesButton.setText(
            String.format(_("dialog_btn_files"), this.files.length));
        if (this.files.length == 0) {
            this.okButton.setEnabled(false);
        } else {
            this.okButton.setEnabled(true);
            this.okButton.grabFocus();
        }
        return;
    }
    
    /**
     * Gets a string for the given key from this resource bundle
     * or one of its parents.  If the key is missing, returns
     * the key itself.
     * 
     * @param key the key for the desired string 
     * @return the string for the given key 
     */
    private String _(String key) {
        try {
            return new String(
                this.l10n.getString(key).getBytes("ISO-8859-1"),
                "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            return this.l10n.getString(key);
        } catch (java.util.MissingResourceException e) {
            return key;
        }
    }
}
